# Fase 1 · Análisis Exploratorio de Datos (EDA)
 
## Objetivo
 
El objetivo de esta fase es realizar una exploración inicial del dataset **DataCo Smart Supply Chain**, el cual simula las operaciones comerciales, logísticas y de cadena de suministro de una organización global. Esta revisión permite comprender la estructura multidimensional de los datos, evaluar su calidad y detectar posibles inconsistencias o sesgos que puedan afectar a las fases posteriores de modelado predictivo y analítica avanzada.
 
En esta etapa nos centramos en:
 
- **Identificar las dimensiones operativas:** Clasificar y entender las variables comerciales (ventas, clientes), logísticas (modos de envío, tiempos de entrega) y de seguridad (estados de órdenes y alertas de fraude).
- **Analizar la consistencia estructural:** Evaluar las dimensiones del dataset, los tipos de datos (fechas, variables categóricas, numéricas) y la coherencia interna entre registros relacionados.
- **Evaluar la calidad de los datos:** Detectar la presencia de valores nulos, registros duplicados, inconsistencias en las coordenadas geográficas y datos potencialmente sensibles o expuestos.
- **Explorar variables críticas del negocio:** Analizar el comportamiento de métricas clave como los días de envío reales frente a los estimados (`SLA`), los patrones de fraude en los métodos de pago y la distribución de ventas por regiones.
- **Anticipar necesidades de ingeniería de características:** Identificar requerimientos específicos de limpieza, normalización de columnas, transformación de variables temporales y preparación de datos para la fase ETL.
 
Este análisis proporciona una base sólida para definir la estrategia de preparación de datos, asegurando un pipeline robusto para la toma de decisiones estratégicas en la cadena de suministro.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import warnings
import sys
import os # Esto sube un nivel en las carpetas para encontrar la raíz del proyecto
sys.path.append(os.path.abspath(os.path.join('..')))
import src.exploration as e

In [2]:
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [3]:
df_data = pd.read_csv("../files/raw/dataco_supply_chain.csv", sep=None, engine="python", encoding="latin-1")

In [4]:
e.info_df(df_data)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 180519
- Columnas: 53

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  str    
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  str    
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  str    
 9   Customer City                  180519 non-null  str    
 10  Customer Country               180519 non-null  str    
 11  Custome

,count,unique,top,freq
Type,180519,4,DEBIT,69295
Delivery Status,180519,4,Late delivery,98977
Category Name,180519,50,Cleats,24551
Customer City,180519,563,Caguas,66770
Customer Country,180519,2,EE. UU.,111146
Customer Email,180519,1,XXXXXXXXX,180519
Customer Fname,180519,782,Mary,65150
Customer Lname,180511,1109,Smith,64104
Customer Password,180519,1,XXXXXXXXX,180519
Customer Segment,180519,3,Consumer,93504



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.50,1.62,0.00,2.00,3.00,5.00,6.00
Days for shipment (scheduled),180519.0,2.93,1.37,0.00,2.00,4.00,4.00,4.00
Benefit per order,180519.0,21.97,104.43,-4274.98,7.00,31.52,64.80,911.80
Sales per customer,180519.0,183.11,120.04,7.49,104.38,163.99,247.40,1939.99
Late_delivery_risk,180519.0,0.55,0.50,0.00,0.00,1.00,1.00,1.00
Category Id,180519.0,31.85,15.64,2.00,18.00,29.00,45.00,76.00
Customer Id,180519.0,6691.38,4162.92,1.00,3258.50,6457.00,9779.00,20757.00
Customer Zipcode,180516.0,35921.13,37542.46,603.00,725.00,19380.00,78207.00,99205.00
Department Id,180519.0,5.44,1.63,2.00,4.00,5.00,7.00,12.00
Latitude,180519.0,29.72,9.81,-33.94,18.27,33.14,39.28,48.78



🔹 Muestra del dataframe:


,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
105311,TRANSFER,2,1,-46.180000,83.970001,Late delivery,1,46,Indoor/Outdoor Games,Westerville,EE. UU.,XXXXXXXXX,Judy,3708,Miller,XXXXXXXXX,Consumer,OH,7774 Broad Crossing,43081.0,7,Fan Shop,40.118351,-82.928108,LATAM,Mexico City,México,3708,6/9/2017 19:25,61025,1014,15.990000,0.16,152667,49.980000,-0.55,2,99.959999,83.970001,-46.180000,Central America,Distrito Federal,PROCESSING,NaN,1014,46,NaN,http://images.acmesports.sports/O%27Brien+Men%...,O'Brien Men's Neoprene Life Vest,49.980000,0,6/11/2017 19:25,First Class
18940,PAYMENT,2,1,144.309998,351.980011,Late delivery,1,45,Fishing,Gardena,EE. UU.,XXXXXXXXX,Mary,9557,Figueroa,XXXXXXXXX,Consumer,CA,7980 Little Rise Village,90247.0,7,Fan Shop,33.891647,-118.299408,Pacific Asia,Seúl,Corea del Sur,9557,12/16/2015 2:29,23916,1004,48.000000,0.12,59858,399.980011,0.41,1,399.980011,351.980011,144.309998,Eastern Asia,Seúl,PENDING_PAYMENT,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,12/18/2015 2:29,First Class
9085,DEBIT,6,4,15.330000,42.220001,Late delivery,1,41,Trade-In,Fairfield,EE. UU.,XXXXXXXXX,Kathleen,2699,Higgins,XXXXXXXXX,Corporate,OH,588 Rocky Quail Round,45014.0,6,Outdoors,39.351505,-84.549995,LATAM,San Salvador,El Salvador,2699,1/26/2017 10:30,51820,917,1.760000,0.04,129468,21.990000,0.36,2,43.980000,42.220001,15.330000,Central America,San Salvador,COMPLETE,NaN,917,41,NaN,http://images.acmesports.sports/Glove+It+Women...,Glove It Women's Mod Oval 3-Zip Carry All Gol,21.990000,0,2/1/2017 10:30,Standard Class
26445,DEBIT,2,1,147.830002,351.980011,Late delivery,1,45,Fishing,Caguas,Puerto Rico,XXXXXXXXX,Mary,8338,Jefferson,XXXXXXXXX,Corporate,PR,7487 Colonial Byway,725.0,7,Fan Shop,18.235300,-66.370552,Europe,Bolton,Reino Unido,8338,8/14/2015 23:41,15482,1004,48.000000,0.12,38695,399.980011,0.42,1,399.980011,351.980011,147.830002,Northern Europe,Inglaterra,COMPLETE,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,8/16/2015 23:41,First Class
108942,PAYMENT,5,4,43.549999,119.970001,Late delivery,1,29,Shop By Sport,Pittsfield,EE. UU.,XXXXXXXXX,Mary,1540,Mckee,XXXXXXXXX,Consumer,MA,1942 Middle Bluff Crescent,1201.0,5,Golf,42.451111,-73.252602,USCA,Springfield,Estados Unidos,1540,5/15/2016 0:21,34254,627,39.990002,0.25,85539,39.990002,0.36,4,159.960007,119.970001,43.549999,East of USA,Ohio,PENDING_PAYMENT,45503.0,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,5/20/2016 0:21,Standard Class


In [5]:
df_data.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [ ]:
df_data[df_data["Customer Id"] == 4995].head() #para corroborar que la latitud y longitud pertenecen al cliente, no a la tienda.

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
5862,CASH,2,1,92.570000,351.980011,Late delivery,1,45,Fishing,Lawrenceville,EE. UU.,XXXXXXXXX,Mary,4995,Smith,XXXXXXXXX,Consumer,GA,1782 High River Rise,30044.0,7,Fan Shop,33.941853,-84.086632,Europe,Leipzig,Alemania,4995,8/10/2015 8:58,15166,1004,48.0,0.12,37934,399.980011,0.26,1,399.980011,351.980011,92.570000,Western Europe,Sajonia,CLOSED,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,8/12/2015 8:58,First Class
15790,CASH,2,1,78.750000,210.000000,Late delivery,1,24,Women's Apparel,Lawrenceville,EE. UU.,XXXXXXXXX,Mary,4995,Smith,XXXXXXXXX,Consumer,GA,1782 High River Rise,30044.0,5,Golf,33.941853,-84.086632,Europe,Leipzig,Alemania,4995,8/10/2015 8:58,15166,502,40.0,0.16,37933,50.000000,0.38,5,250.000000,210.000000,78.750000,Western Europe,Sajonia,CLOSED,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,8/12/2015 8:58,First Class
54959,DEBIT,3,4,70.500000,149.990005,Advance shipping,0,48,Water Sports,Lawrenceville,EE. UU.,XXXXXXXXX,Mary,4995,Smith,XXXXXXXXX,Consumer,GA,1782 High River Rise,30044.0,7,Fan Shop,33.941853,-84.086632,LATAM,Bogotá,Colombia,4995,5/9/2017 17:53,58897,1073,50.0,0.25,147430,199.990005,0.47,1,199.990005,149.990005,70.500000,South America,Bogotá,COMPLETE,NaN,1073,48,NaN,http://images.acmesports.sports/Pelican+Sunstr...,Pelican Sunstream 100 Kayak,199.990005,0,5/12/2017 17:53,Standard Class
78943,PAYMENT,3,4,-148.789993,193.990005,Advance shipping,0,48,Water Sports,Lawrenceville,EE. UU.,XXXXXXXXX,Mary,4995,Smith,XXXXXXXXX,Consumer,GA,1782 High River Rise,30044.0,7,Fan Shop,33.941853,-84.086632,USCA,Oceanside,Estados Unidos,4995,6/18/2016 6:57,36602,1073,6.0,0.03,91369,199.990005,-0.77,1,199.990005,193.990005,-148.789993,East of USA,Nueva York,PENDING_PAYMENT,11572.0,1073,48,NaN,http://images.acmesports.sports/Pelican+Sunstr...,Pelican Sunstream 100 Kayak,199.990005,0,6/21/2016 6:57,Standard Class
84158,DEBIT,3,4,-125.059998,187.490005,Advance shipping,0,41,Trade-In,Lawrenceville,EE. UU.,XXXXXXXXX,Mary,4995,Smith,XXXXXXXXX,Consumer,GA,1782 High River Rise,30044.0,6,Outdoors,33.941853,-84.086632,LATAM,Bogotá,Colombia,4995,5/9/2017 17:53,58897,607,62.5,0.25,147427,249.990005,-0.67,1,249.990005,187.490005,-125.059998,South America,Bogotá,COMPLETE,NaN,607,41,NaN,http://images.acmesports.sports/Garmin+Approac...,Garmin Approach S3 Golf GPS Watch,249.990005,0,5/12/2017 17:53,Standard Class


In [ ]:
df_data.groupby("Department Id")["Order Id"].value_counts().sort_values(ascending=False).head(10)

Department Id  Order Id
5              31242       5
7              58549       5
               5443        5
               25752       5
               26862       5
               55052       5
               56068       5
               52950       5
               33118       5
               25135       5
Name: count, dtype: int64

In [34]:
df_data[df_data["Department Id"] == 5].sample(5) #para corroborar que la latitud y longitud pertenecen al cliente, no a la tienda.

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
97758,DEBIT,2,2,12.35,47.500000,Shipping on time,0,24,Women's Apparel,Union City,EE. UU.,XXXXXXXXX,Jerry,10627,Smith,XXXXXXXXX,Home Office,NJ,3378 Old Drive,7087.0,5,Golf,40.761524,-74.038269,Pacific Asia,Mandurah,Australia,10627,12/23/2015 4:24,24401,502,2.500000,0.05,61037,50.000000,0.26,1,50.000000,47.500000,12.35,Oceania,Australia Occidental,COMPLETE,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,12/25/2015 4:24,Second Class
128233,DEBIT,3,4,27.84,87.000000,Advance shipping,0,24,Women's Apparel,New Brunswick,EE. UU.,XXXXXXXXX,Mary,8726,Diaz,XXXXXXXXX,Consumer,NJ,3305 Umber Mews,8901.0,5,Golf,40.492355,-74.443733,USCA,New York City,Estados Unidos,8726,6/10/2016 8:01,36057,502,13.000000,0.13,90029,50.000000,0.32,2,100.000000,87.000000,27.84,East of USA,Nueva York,COMPLETE,10009.0,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,6/13/2016 8:01,Standard Class
122287,TRANSFER,1,0,34.84,105.570000,Late delivery,1,29,Shop By Sport,Caguas,Puerto Rico,XXXXXXXXX,Cheryl,6744,Medina,XXXXXXXXX,Corporate,PR,779 Velvet Blossom Key,725.0,5,Golf,18.213333,-66.370575,LATAM,João Pessoa,Brasil,6744,1/20/2017 14:00,51419,627,14.400000,0.12,128500,39.990002,0.33,3,119.970001,105.570000,34.84,South America,Paraíba,PENDING,NaN,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,1/21/2017 2:00,Same Day
150109,DEBIT,4,4,-103.18,140.759995,Shipping on time,0,29,Shop By Sport,Caguas,Puerto Rico,XXXXXXXXX,Kelly,6355,Bradley,XXXXXXXXX,Home Office,PR,7802 Grand Spring Chase,725.0,5,Golf,18.283676,-66.370522,LATAM,Arraiján,Panamá,6355,3/8/2015 14:16,4563,627,19.200001,0.12,11432,39.990002,-0.73,4,159.960007,140.759995,-103.18,Central America,Panamá,COMPLETE,NaN,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,3/12/2015 14:16,Standard Class
121962,TRANSFER,2,4,3.75,33.189999,Advance shipping,0,29,Shop By Sport,Philadelphia,EE. UU.,XXXXXXXXX,Arthur,11330,Johnson,XXXXXXXXX,Consumer,PA,5443 Old Timber Expressway,19143.0,5,Golf,39.946629,-75.218224,LATAM,Guarujá,Brasil,11330,1/9/2015 23:27,616,627,6.800000,0.17,1505,39.990002,0.11,1,39.990002,33.189999,3.75,South America,São Paulo,PROCESSING,NaN,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,1/11/2015 23:27,Standard Class


In [27]:
df_data.groupby("Customer Id")["Order Id"].value_counts().sort_values(ascending=False).head(10)

Customer Id  Order Id
2            57963       5
4995         58897       5
4996         36451       5
             6717        5
4997         63047       5
4987         45977       5
             30412       5
4989         11862       5
12425        48314       5
12426        32304       5
Name: count, dtype: int64

In [6]:
df_data["shipping date (DateOrders)"].value_counts()

shipping date (DateOrders)
5/9/2015 18:02      10
9/27/2015 20:28     10
9/24/2015 23:48     10
8/1/2015 18:37      10
11/15/2016 13:45    10
                    ..
1/27/2016 18:59      1
1/28/2016 19:25      1
1/26/2016 6:49       1
1/23/2016 7:10       1
1/21/2016 5:56       1
Name: count, Length: 63701, dtype: int64

In [7]:
df_data["Shipping Mode"].value_counts()

Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

In [8]:
df_data["Customer Country"].value_counts()

Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

In [9]:
df_data["Customer State"].unique()

<StringArray>
[   'PR',    'CA',    'NY',    'FL',    'MA',    'IL',    'MT',    'PA',
    'MI',    'TX',    'DE',    'GA',    'MD',    'OH',    'HI',    'NJ',
    'WI',    'AZ',    'CO',    'MN',    'NC',    'NM',    'OR',    'SC',
    'VA',    'UT',    'WA',    'KY',    'WV',    'RI',    'CT',    'LA',
    'TN',    'DC',    'ND',    'MO',    'IN',    'ID',    'NV',    'KS',
    'AR',    'OK',    'AL',    'IA', '95758', '91732']
Length: 46, dtype: str

In [10]:
df_data["Order Status"].unique()

<StringArray>
[       'COMPLETE',         'PENDING',          'CLOSED', 'PENDING_PAYMENT',
        'CANCELED',      'PROCESSING', 'SUSPECTED_FRAUD',         'ON_HOLD',
  'PAYMENT_REVIEW']
Length: 9, dtype: str

In [11]:
df_data["Order Region"].unique()

<StringArray>
[ 'Southeast Asia',      'South Asia',         'Oceania',    'Eastern Asia',
       'West Asia',    'West of USA ',      'US Center ',     'West Africa',
  'Central Africa',    'North Africa',  'Western Europe', 'Northern Europe',
 'Central America',       'Caribbean',   'South America',     'East Africa',
 'Southern Europe',     'East of USA',          'Canada', 'Southern Africa',
    'Central Asia',  'Eastern Europe',  'South of  USA ']
Length: 23, dtype: str

In [12]:
df_data["Market"].unique()

<StringArray>
['Pacific Asia', 'USCA', 'Africa', 'Europe', 'LATAM']
Length: 5, dtype: str

In [13]:
df_data["Category Name"].unique()

<StringArray>
[      'Sporting Goods',               'Cleats',        'Shop By Sport',
      'Women's Apparel',          'Electronics',         'Boxing & MMA',
     'Cardio Equipment',             'Trade-In',     'Kids' Golf Clubs',
   'Hunting & Shooting',  'Baseball & Softball',       'Men's Footwear',
     'Camping & Hiking', 'Consumer Electronics',             'Cameras ',
            'Computers',           'Basketball',               'Soccer',
       'Girls' Apparel',          'Accessories',     'Women's Clothing',
               'Crafts',       'Men's Clothing',     'Tennis & Racquet',
  'Fitness Accessories',      'As Seen on  TV!',           'Golf Balls',
    'Strength Training',  'Children's Clothing',             'Lacrosse',
                'Baby ',              'Fishing',               'Books ',
                 'DVDs',                 'CDs ',               'Garden',
               'Hockey',         'Pet Supplies',    'Health and Beauty',
                'Music',          'Vi

In [14]:
df_data["Customer Segment"].unique()

<StringArray>
['Consumer', 'Home Office', 'Corporate']
Length: 3, dtype: str

In [15]:
df_data.groupby("Customer State")["Order Zipcode"].unique()

Customer State
91732                                                [nan]
95758                                                [nan]
AL                [nan, 60462.0, 98105.0, 7960.0, 22980.0]
AR               [nan, 43130.0, 90032.0, 39212.0, 43302.0]
AZ       [nan, 10035.0, 32216.0, 43229.0, 46368.0, 8001...
CA       [nan, 95123.0, 19134.0, 31907.0, 77041.0, 2230...
CO       [nan, 26003.0, 38109.0, 92024.0, 19134.0, 1002...
CT       [nan, 89115.0, 19140.0, 95051.0, 90049.0, 7501...
DC       [nan, 73120.0, 31907.0, 48227.0, 47201.0, 9203...
DE       [nan, 43055.0, 60505.0, 90049.0, 90032.0, 7050...
FL       [nan, 38109.0, 72401.0, 60610.0, 6708.0, 43017...
GA       [nan, 10035.0, 97206.0, 98661.0, 87124.0, 7570...
HI       [60016.0, nan, 75034.0, 13440.0, 89031.0, 1000...
IA               [nan, 10011.0, 85023.0, 48911.0, 78041.0]
ID               [nan, 38301.0, 48234.0, 20735.0, 19711.0]
IL       [nan, 13021.0, 79762.0, 32216.0, 49201.0, 9811...
IN       [nan, 37042.0, 60623.0, 94110.0,

In [16]:
df_token = pd.read_csv("../files/raw/tokenized_access_logs.csv")

In [17]:
e.info_df(df_token)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 469977
- Columnas: 8

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 469977 entries, 0 to 469976
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype
---  ------      --------------   -----
 0   Product     469977 non-null  str  
 1   Category    469977 non-null  str  
 2   Date        469977 non-null  str  
 3   Month       469977 non-null  str  
 4   Hour        469977 non-null  int64
 5   Department  469977 non-null  str  
 6   ip          469977 non-null  str  
 7   url         469977 non-null  str  
dtypes: int64(1), str(7)
memory usage: 28.7 MB

🔹 Nulos:
Product       0.0
Category      0.0
Date          0.0
Month         0.0
Hour          0.0
Department    0.0
ip            0.0
url           0.0
dtype: float64

🔹 Duplicados:
Total duplicados: 3249

🔹 Descriptivo variables cualitativas:


,count,unique,top,freq
Product,469977,76,Perfect Fitness Perfect Rip Deck,27878
Category,469977,33,cleats,27878
Date,469977,160815,9/14/2017 20:34,156
Month,469977,5,Sep,137238
Department,469977,6,outdoors,79926
ip,469977,3340,157.21.93.193,566
url,469977,152,/department/apparel/category/cleats/product/Pe...,20258



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Hour,469977.0,14.59,5.57,0.0,10.0,15.0,20.0,23.0



🔹 Muestra del dataframe:


,Product,Category,Date,Month,Hour,Department,ip,url
285196,Ogio Race Golf Shoes,golf bags & carts,1/3/2018 14:08,Jan,14,outdoors,105.130.122.91,/department/outdoors/category/golf%20bags%20&%...
7227,Bushnell Pro X7 Jolt Slope Rangefinder,kids' golf clubs,9/2/2017 14:42,Sep,14,outdoors,104.82.47.72,/department/outdoors/category/kids'%20golf%20c...
94847,SOLE E25 Elliptical,basketball,10/25/2017 7:30,Oct,7,fitness,100.161.130.203,/department/fitness/category/basketball/produc...
23659,Nike Men's Dri-FIT Victory Golf Polo,women's apparel,9/4/2017 7:26,Sep,7,golf,150.40.221.1,/department/golf/category/women's%20apparel/pr...
138771,Nike Men's Free TR 5.0 TB Training Shoe,as seen on tv!,12/17/2017 14:15,Dec,14,footwear,119.244.8.90,/department/footwear/category/as%20seen%20on%2...


In [18]:
df_token["Month"].unique()

<StringArray>
['Sep', 'Oct', 'Dec', 'Jan', 'Nov']
Length: 5, dtype: str

# Conclusiones finales del EDA · DataCo
 
- **Calidad de Datos y Gestión de Nulos:** La integridad general del dataset es excelente en la mayoría de sus variables transaccionales y financieras. La excepción principal se concentra en la columna `Order Zipcode`, la cual presenta valores nulos que requieren una estrategia de imputación o limpieza en la fase ETL, dado que pueden afectar el análisis geoespacial del destino de las órdenes.
 
- **Duplicados y Estructura Lógica:** Las repeticiones detectadas en los registros no constituyen errores de duplicidad aleatoria; se justifican plenamente por la estructura lógica y granular del dataset, donde un mismo número de orden (`Order Id`) puede contener múltiples ítems de productos individuales o registrar distintas actualizaciones en el historial de envío.
 
- **Políticas de Privacidad y Reducción de Dimensiones (Eliminación):** Con el fin de optimizar el rendimiento del almacenamiento en el ecosistema de Big Data y garantizar el cumplimiento estricto de las normativas de protección de datos (debido a la naturaleza sintética/expuesta de cierta información), se determina la eliminación definitiva de las siguientes columnas no relevantes o sensibles:
  - `Customer Email` (Privacidad)
  - `Customer Password` (Seguridad)
  - `Product Description` (Columna redundante o vacía en el catálogo de productos)
  - `Order State` (Duplicidad con las métricas regionales o de envío)
 
- **Estandarización y Corrección de Anomalías:** Se detectaron inconsistencias críticas en la columna de ubicación del cliente (`Customer State`). Mientras que la mayoría de los registros adoptan el formato estándar de código de estado de dos letras (ej. NY, TX), se identificaron códigos postales numéricos (como *95758* y *91732*) en dicho campo. Como acción obligatoria de limpieza, se plantea mapear y reemplazar estos números por el estado correcto al que pertenecen (`CA`).
 
- **Optimización de Estructura y Formato de Columnas:** Dado que las columnas ya se encuentran en el idioma inglés, la transformación se centrará en estandarizar su formato de texto. Se aplicará una conversión estricta a minúsculas y se reemplazarán los espacios por guiones bajos (formato `snake_case`) para homogeneizar los nombres de las variables (ej. de `Order Id` a `order_id`). Asimismo, se transformarán las variables temporales desde texto plano a formato `datetime` nativo.
 
- **Integración y Complementariedad:** Las tablas analizadas demuestran ser altamente complementarias para construir un modelo en estrella o copo de nieve, permitiendo cruzar de manera directa las dimensiones de Clientes, Productos y Geografía con la tabla de hechos de Órdenes y Envíos.
 
---
 
Esta fase concluye con éxito, dejando una hoja de ruta clara, justificada y accionable para la siguiente etapa del pipeline: **Limpieza, Transformación y Carga (ETL)**.